File: nudgly/viewmodels/main_viewmodel.py
from PySide6.QtCore import QObject, Signal, Slot, Property

from nudgly.models.conversation_model import ConversationModel
from nudgly.services.llm_service import LLMService


class MainViewModel(QObject):
    answerResultChanged = Signal()
    answersWindowVisibilityChanged = Signal()
    isThinkingChanged = Signal()

    def __init__(self, conversation_model: ConversationModel):
        super().__init__()

        self._conversation_model = conversation_model

        # --- Instantiate the service ---
        self._llm_service = LLMService(self)

        # --- Connect service signals to local slots ---
        self._llm_service.answerGenerated.connect(self._on_answer_received)
        self._llm_service.errorOccurred.connect(self._on_api_error)

        # --- Properties ---
        self._answerResult: str = ""
        self._answersWindowVisible: bool = False
        self._isThinking: bool = False

    # --- Public Slots (callable from UI) ---
    @Slot()
    def askAi(self):
        """
        This slot is now non-blocking. It starts the AI task and returns immediately.
        """
        self.isThinking = True
        # Call the asynchronous method from the service
        self._llm_service.generate_answer(self._conversation_model)

    @Slot()
    def toggleAnswersWindowVisibility(self):
        self.answersWindowVisible = not self.answersWindowVisible

    @Slot()
    def openSettings(self):
        print("Settings window requested")

    # --- Private Slots (for handling service signals) ---
    @Slot(str)
    def _on_answer_received(self, assistant_response: str):
        """Handles the successful result from the LLMService."""
        self.answerResult = assistant_response
        self.isThinking = False

    @Slot(str)
    def _on_api_error(self, error_message: str):
        """Handles errors from the LLMService."""
        # You might want to format this for the user
        self.answerResult = f"An error occurred: {error_message}"
        self.isThinking = False

    # --- Property Definitions (unchanged) ---
    def get_answer_result(self):
        return self._answerResult

    def set_answer_result(self, value):
        if self._answerResult != value:
            self._answerResult = value
            self.answersWindowVisible = True
            self.answerResultChanged.emit()

    answerResult = Property(str, get_answer_result, set_answer_result, notify=answerResultChanged)

    def get_answers_window_visible(self):
        return self._answersWindowVisible

    def set_answers_window_visible(self, value):
        if self._answersWindowVisible != value:
            self._answersWindowVisible = value
            self.answersWindowVisibilityChanged.emit()

    answersWindowVisible = Property(bool, get_answers_window_visible, set_answers_window_visible,
                                    notify=answersWindowVisibilityChanged)

    def get_is_thinking(self):
        return self._isThinking

    def set_is_thinking(self, value):
        if self._isThinking != value:
            self._isThinking = value
            self.isThinkingChanged.emit()

    isThinking = Property(bool, get_is_thinking, set_is_thinking, notify=isThinkingChanged)

File: src/ViewModels/MainViewModel.cpp
#include "ViewModels/MainViewModel.h"

#include <QDebug>

MainViewModel::MainViewModel(Models::ConversationModel::ConversationModel* conversationModel, QObject* parent)
    : QObject(parent),
      m_conversationModel(conversationModel),
      m_llmService(new Services::LLMService(this)) // Instantiate the service
{
    // Connect service signals to our private handler slots
    connect(m_llmService, &Services::LLMService::answerChunkReceived, this, &MainViewModel::onAnswerChunkReceived);
    connect(m_llmService, &Services::LLMService::answerGenerated, this, &MainViewModel::onAnswerFinished);
    connect(m_llmService, &Services::LLMService::errorOccurred, this, &MainViewModel::onApiError);
}

// --- Public Slots ---

void MainViewModel::askAi()
{
    // This is non-blocking. It starts the task and returns immediately.
    setAnswerResult("");
    setIsThinking(true);
    m_llmService->generateAnswer(m_conversationModel);
}

void MainViewModel::toggleAnswersWindowVisibility()
{
    setAnswersWindowVisible(!m_answersWindowVisible);
}

void MainViewModel::openSettings()
{
    qDebug() << "Settings window requested";
}

// --- Private Slots ---

void MainViewModel::onAnswerChunkReceived(const QString& chunk)
{
    // Append the incoming text chunk to the current answer result.
    // The setter will emit the changed signal, updating the UI in real-time.
    setAnswerResult(m_answerResult + chunk);
}

void MainViewModel::onAnswerFinished(const QString& fullAnswer)
{
    // The stream has finished. Finalize the state.
    // This ensures the final text is consistent and, most importantly,
    // it updates the thinking state to stop any loading indicators in the UI.
    setAnswerResult(fullAnswer);
    setIsThinking(false);
}

void MainViewModel::onApiError(const QString& errorMessage)
{
    // Format the error message for the user
    setAnswerResult(QString("An error occurred: %1").arg(errorMessage));
    setIsThinking(false);
}

// --- Property Accessors ---

QString MainViewModel::answerResult() const
{
    return m_answerResult;
}

void MainViewModel::setAnswerResult(const QString& answerResult)
{
    if (m_answerResult != answerResult)
    {
        m_answerResult = answerResult;
        emit answerResultChanged();

        // Side-effect: show the answers window when a new result arrives
        if (answerResult.isEmpty())
        {
            setAnswersWindowVisible(true);
        }
    }
}

bool MainViewModel::answersWindowVisible() const
{
    return m_answersWindowVisible;
}

void MainViewModel::setAnswersWindowVisible(const bool visible)
{
    if (m_answersWindowVisible != visible)
    {
        m_answersWindowVisible = visible;
        emit answersWindowVisibilityChanged();
    }
}

bool MainViewModel::isThinking() const
{
    return m_isThinking;
}

void MainViewModel::setIsThinking(const bool thinking)
{
    if (m_isThinking != thinking)
    {
        m_isThinking = thinking;
        emit isThinkingChanged();
    }
}
